### PR TITLE
Issue 4459: [r0.6] Upgrading bookkeeper version to 4.8.2

### LINKED ISSUE #4459: Pravega Bookkeeper docker image has an old version of Bookkeeper server
**Problem description**
Pravega Bookkeeper image currently uses an old version of Bookkeeper: `4.7.3`. Many defects were fixed in later versions of Bookeeper, including but not limited to the `FileNotFoundException` reported in issue #4270 .

**Problem location**
https://github.com/pravega/pravega/blob/5d0176839755e82f64329938466099e25088c3ad/docker/bookkeeper/Dockerfile#L12

**Suggestions for an improvement**
Upgrade the Bookkeeper version

### PR DESCRIPTION
**Pull request by SrishT**
Signed-off-by: SrishT <Srishti.Thakkar@dell.com>

**Change log description**  
Upgrade the Bookeeper server from `4.7.3` to `4.8.2`.

**Purpose of the change**  
Resolves #4459 

**What the code does**  
Modifies the Dockerfile and entrypoint to work with Bookeeper `4.8.2`. It does the following:

1. The default 4.8.2 bookkeeper start up scripts uses `ZooKeeperMain` to create/modify nodes in ZK. This throws
```
/opt/bookkeeper/bin/bookkeeper org.apache.zookeeper.ZooKeeperMain -server
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 3
```
Due to this, we ended up in using `zk-shell` on similar lines as Apache Bookeeper PR https://github.com/apache/bookkeeper/pull/1666.

2. The `4.8.2` script exits if `BK_STREAM_STORAGE_ROOT_PATH` is not set. Also, a zk node in this path is expected to be present if a bookie is up and running. (This is not the case with the default bk 4.8.2 image)

3. `/opt/bookkeeper/bin/bkctl --service-uri` this option used if Bookkeeper image did not work. `java.lang.IllegalArgumentException: No service URI` is observed. It is similar to issues observed in apache/bookkeeper#2112.

4. It has a workaround for creating ephmeral node using `zk-shell`. The exit status returned by the `/opt/bookkeeper/bin/bookkeeper org.apache.zookeeper.ZooKeeperMain` used in the base image is always 0, even in the case the node exists, causing unnecessary bookie restart.

**How to verify it**  
* A successful system test run
* A successful longevity test run
* Manual testing of at least the following scenarios:
  * Upgrade
  * Failure of Bookkeeper instances in a cluster

### CODE REVIEW

**Comment by eolivelli** [comment]
Why are you upgrading to 4.82 and not to 4.9.2?

File: docker/bookkeeper/Dockerfile [hunk 1]
```diff
@@ -7,31 +7,35 @@
...
-ARG GPG_KEY=FD74402C
 
 RUN set -x \
-    && yum install -y iproute \
+    && yum install -y iproute wget \
     && cd /opt \
-    && curl -O "https://archive.apache.org/dist/bookkeeper/bookkeeper-${BK_VERSION}/${DISTRO_NAME}.tar.gz" \
-    && curl -O "https://archive.apache.org/dist/bookkeeper/bookkeeper-${BK_VERSION}/${DISTRO_NAME}.tar.gz.asc" \
-    && curl -O "https://archive.apache.org/dist/bookkeeper/bookkeeper-${BK_VERSION}/${DISTRO_NAME}.tar.gz.sha512" \
+    && wget -q "https://archive.apache.org/dist/bookkeeper/bookkeeper-${BK_VERSION}/${DISTRO_NAME}.tar.gz" \
```
Comment (fpj) [inline docker/bookkeeper/Dockerfile]: Not a big deal, but it seems to be out of scope to change from `curl` to `wget`.

File: docker/bookkeeper/entrypoint.sh [hunk 2]
```diff
@@ -48,31 +47,143 @@ export BK_tlsTrustStoreType=JKS
...
+
+configure_bk() {
+    # We need to update the metadata endpoint and Bookie ID before attempting to delete the cookie
+    sed -i "s|.*metadataServiceUri=.*\$|metadataServiceUri=${BK_metadataServiceUri}|" /opt/bookkeeper/conf/bk_server.conf
+    if [ ! -z "$BK_useHostNameAsBookieID" ]; then
+      sed -i "s|.*useHostNameAsBookieID=.*\$|useHostNameAsBookieID=${BK_useHostNameAsBookieID}|" ${BK_HOME}/conf/bk_server.conf
+    fi
+}
+
+fix_bk_ipv6_check() {
```
Comment (fpj) [inline docker/bookkeeper/entrypoint.sh]: This feels a bit hacky. Could we not exit on error and check the error code instead? If it is 255 with that message, then we ignore rather than exit.

File: docker/bookkeeper/entrypoint.sh [hunk 3]
```diff
@@ -48,31 +47,143 @@ export BK_tlsTrustStoreType=JKS
...
+            if [ $? -eq 0 ]; then
+                echo "initnewcluster operation succeeded"
+            else
+                echo "initnewcluster operation failed. Please check the reason."
+                echo "Exit status of initnewcluster"
+                echo $?
+                exit
+            fi
+        else
+            echo "Others may be initializing the cluster at the same time."
```
Comment (fpj) [inline docker/bookkeeper/entrypoint.sh]: What others can be initializing the cluster at the same time?

File: docker/bookkeeper/entrypoint.sh [hunk 4]
```diff
@@ -48,31 +47,143 @@ export BK_tlsTrustStoreType=JKS
...
+                    break
+                else
+                    echo "Waited $tenSeconds * 10 seconds. Continue waiting."
+                    (( tenSeconds++ ))
+                    continue
+                fi
+            done
+
+            if [ ${tenSeconds} -eq 1000 ]; then
+                echo "Waited 1000 seconds for bookkeeper cluster to initialize, but to no . Something is wrong, please check"
```
Comment (fpj) [inline docker/bookkeeper/entrypoint.sh]: This error message is incomplete, please revisit it.

File: docker/bookkeeper/entrypoint.sh [hunk 5]
```diff
@@ -48,31 +47,143 @@ export BK_tlsTrustStoreType=JKS
...
+            else
+                echo "initnewcluster operation failed. Please check the reason."
+                echo "Exit status of initnewcluster"
+                echo $?
+                exit
+            fi
+        else
+            echo "Others may be initializing the cluster at the same time."
+            tenSeconds=1
+            while [ ${tenSeconds} -lt 100 ]
```
Comment (fpj) [inline docker/bookkeeper/entrypoint.sh]: How are we choosing the max time to wait? This seems to be waiting for 16-17 minutes max.

File: docker/bookkeeper/entrypoint.sh [hunk 6]
```diff
@@ -48,31 +47,143 @@ export BK_tlsTrustStoreType=JKS
...
+
+create_zk_root() {
+    if [ "x${BK_CLUSTER_ROOT_PATH}" != "x" ]; then
+        echo "Creating the zk root dir '${BK_CLUSTER_ROOT_PATH}' at '${BK_zkServers}'"
+        zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH} '' false false true" ${BK_zkServers}
+    fi
+}
+
+configure_bk() {
+    # We need to update the metadata endpoint and Bookie ID before attempting to delete the cookie
```
Comment (fpj) [inline docker/bookkeeper/entrypoint.sh]: Could you add to the comment to elaborate on what's really happening? Explain what a cookie is and why the metadata service endpoint needs to be updated. Otherwise, it is difficult to understand what it going on.

File: docker/bookkeeper/entrypoint.sh [hunk 7]
```diff
@@ -48,31 +47,143 @@ export BK_tlsTrustStoreType=JKS
...
+   #    `sysctl: cannot stat /proc/sys/net/ipv6/bindv6only: No such file or directory`
+   #
+   # If the `common.sh` file is sourced in a script that sets "exit on error" as true (set -e), the script shall return
+   # an error code `255`. Since, the `bookkeeper/bin/bookeeper` shell script sources the `common.sh` file, and it sets
+   # "exit on error" as true, we encounter this error when it is invoked. To avoid that error, here we modify the
+   # `common.sh` file to execute `/sbin/sysctl -n net.ipv6.bindv6only` only if file `/proc/sys/net/ipv6/bindv6only`
+   # is available.
+
+   # Replace the 22nd line with "if [ -f /sbin/sysctl ] && [ -f /proc/sys/net/ipv6/bindv6only ]; then".
+   sed -i "22s|.*|if [ -f /sbin/sysctl ] \&\& [ -f /proc/sys/net/ipv6/bindv6only ]; then|" ${BK_HOME}/bin/common.sh
```
Comment (fpj) [inline docker/bookkeeper/entrypoint.sh]: it looks like we can propose a change upstream to Apache BookKeeper rather than making patches like this. Are you considering doing it?

File: docker/bookkeeper/entrypoint.sh [hunk 8]
```diff
@@ -48,31 +47,143 @@ export BK_tlsTrustStoreType=JKS
...
+}
+
+initialize_cluster() {
+    set +e
+
+    zk-shell --run-once "ls ${BK_zkLedgersRootPath}/available/readonly" ${BK_zkServers}
+    if [ $? -eq 0 ]; then
```
Comment (fpj) [inline docker/bookkeeper/entrypoint.sh]: The description of the PR says:

```
The exit status returned by the zk-shell is always 0, even in the case the node exists, causing unnecessary bookie restart.)
```

If it is always zero, then I wonder if the behavior of this if-else block is correct.

File: docker/bookkeeper/Dockerfile (hunk 1 above)
Comment (ravisharda) [reply docker/bookkeeper/Dockerfile]: The base Bookkeeper image uses `wget` now, so this was modified too. 

https://github.com/apache/bookkeeper/blob/branch-4.8/docker/Dockerfile

File: docker/bookkeeper/entrypoint.sh (hunk 2 above)
Comment (ravisharda) [reply docker/bookkeeper/entrypoint.sh]: Yes, it is indeed hacky. The problem is that one of the internal Bookkeeper script invokes the `${SCRIPTS_DIR}/common.sh` script too, so we must modify the `common.sh` file itself. 

Eventually, this change would not be required once the [Bookkeeper PR](https://github.com/apache/bookkeeper/pull/2219) we have raised is merged and we modify our image to leverage the modified image.

File: docker/bookkeeper/entrypoint.sh (hunk 4 above)
Comment (ravisharda) [reply docker/bookkeeper/entrypoint.sh]: @SrishT, 
It looks like not all of the changes made in the source [PR 4420](https://github.com/pravega/pravega/pull/4420/) (for the upgrade to `4.9.2`)  have made it to this PR. This error message was modified in the source PR: https://github.com/pravega/pravega/blob/5c4c77fb7febc5e548807d2c8750bc87c368d6db/docker/bookkeeper/entrypoint.sh#L149. 

Could you please integrate the latest updates made in the source PR into this PR?

File: docker/bookkeeper/entrypoint.sh (hunk 3 above)
Comment (ravisharda) [reply docker/bookkeeper/entrypoint.sh]: It is other "instances" of Bookkeeper containers.

File: docker/bookkeeper/entrypoint.sh (hunk 7 above)
Comment (ravisharda) [reply docker/bookkeeper/entrypoint.sh]: Yes, there is an open PR for `apache/bookkeeper`: https://github.com/apache/bookkeeper/pull/2219.

File: docker/bookkeeper/entrypoint.sh (hunk 8 above)
Comment (ravisharda) [reply docker/bookkeeper/entrypoint.sh]: Good catch, @fpj ! :) The description was erroneous. I have updated the description in this (as well as the [source PR](https://github.com/pravega/pravega/pull/4420/)).

**Comment by ravisharda** [comment]
PR #4420 now targets BookKeeper 4.8.2 instead of 4.9.2. Accordingly, this PR is now redundant. So, closing this PR.

**Comment by eolivelli** [comment]
@ravisharda why are we going to 4.8.2?

Latest BK release is 4.10, latest stable is 4.9.2.
4.8 branch is almost dead

**Comment by ravisharda** [comment]
> @ravisharda why are we going to 4.8.2?
> 
> Latest BK release is 4.10, latest stable is 4.9.2.
> 4.8 branch is almost dead

@eolivelli upgrading to `4.8.2` is a temporary/tactical thing, and we'll upgrade Bookkeeper to higher version shortly later after we have had the chance to verify that everything works and performs as expected.

**Comment by eolivelli** [comment]
I see.
Thanks
